src/Hack.sol

```diff
@@ -626,55 +629,74 @@ contract WinnerTakesAll {
+    }
+
     /// @dev Set the reward at a specific round.
-    /// @param _roundIndex The index of the round concerned by the reward.
-    function setRewardsAtRound(uint256 _roundIndex) external payable onlyOwner {
-        require(rounds[_roundIndex].rewards == 0);
-        rounds[_roundIndex].rewards = msg.value;
-    }
-
-    /// @dev Allows the participation of a set of addresses for a specific round.
-    /// @param _roundIndex The index of the round concerned.
-    /// @param _recipients The set of addresses allowed to participate.
-    function setRewardsAtRoundFor(uint256 _roundIndex, address[] calldata _recipients) external onlyOwner {
-        for (uint256 i; i < _recipients.length; i++) {
-            rounds[_roundIndex].isAllowed[_recipients[i]] = true;
-        }
+    /// @param lotteryIndex The index of the round concerned by the reward.
+    function setRewardsAtRound(uint256 lotteryIndex) external payable onlyOwner {
+        require(lotteries[lotteryIndex].rewards == 0);
+        lotteries[lotteryIndex].rewards = msg.value;
     }
 
-    /// @dev Checks if an address can participate in this round.
-    /// @param _roundIndex The index of the round to be checked.
-    /// @param _recipient The address whose authorization is to be checked.
-    function isAllowedAt(uint256 _roundIndex, address _recipient) external view returns (bool) {
-        return rounds[_roundIndex].isAllowed[_recipient];
+    /// @dev Set the winning number. It is chosen randomly off-chain by the trusted owner.
+    /// @param lotteryIndex The index of the round concerned.
+    /// @param winningNumber The winning number of the lottery.
+    function setWinningNumberAtRound(uint256 lotteryIndex, uint256 winningNumber) external onlyOwner {
+        require(winningNumber <= lotteries[lotteryIndex].ticketNumber, "Incorrect winning ticket");
+        require(winningNumber != 0, "Incorrect winning ticket");
+        lotteries[lotteryIndex].winningNumber = winningNumber;
     }
 
     /// @dev Withdraws rewards of a round.
-    /// @param _roundIndex The index of the round concerned.
-    function withdrawRewards(uint256 _roundIndex) external {
-        require(rounds[_roundIndex].isAllowed[msg.sender]);
-        uint256 amount = rounds[_roundIndex].rewards;
-        rounds[_roundIndex].rewards = 0;
-        (bool success,) = msg.sender.call{value: amount}("");
-        require(success, "Transfer failed");
+    /// @param lotteryIndex The index of the round concerned.
+    function withdrawRewards(uint256 lotteryIndex) external {
+        uint256 winningTicket = lotteries[lotteryIndex].winningNumber;
+        require(winningTicket != 0, "Incorrect winning ticket");
+
+        uint256[] memory numbers = lotteries[lotteryIndex].ticketDistribution[msg.sender];
+
+        uint256 amount = lotteries[lotteryIndex].rewards;
+        lotteries[lotteryIndex].rewards = 0;
+
+        for (uint256 i = 0; i < numbers.length; i++) {
```

Comment:
hmm this can lead to another attack where you put too many variables so that this loop could not fit into one block...

The caller could just pass the correct `winningTicket` and we send rewards to her?

Reply:
Agree with you but to buy a ticket, you need to spend one ether. And you can block the rewards only if you win which is a bit stupid. So to block the contract, you need to win and spend a lot of money, it is not so important in my opinion. Curious to have your opinion on this

Reply:
Yes that's true